### PR TITLE
Add constraints

### PR DESCRIPTION
**Pull request by Pratyush**

### CODE REVIEW

**Comment by weikengchen** [comment]
Done. Please take a look. If okay, I will likely use this as a start for Gm17.

File: Cargo.toml [hunk 1]
```diff
@@ -22,23 +22,27 @@ ark-poly = { git = "https://github.com/arkworks-rs/algebra", default-features =
 ark-std = { git = "https://github.com/arkworks-rs/utils", default-features = false }
 bench-utils = { git = "https://github.com/arkworks-rs/utils" }
 ark-relations = { git = "https://github.com/arkworks-rs/snark", default-features = false }
-ark-snark = { git = "https://github.com/arkworks-rs/snark", default-features = false }
+ark-crypto-primitives = { git = "https://github.com/arkworks-rs/crypto-primitives.git", branch = "main", default-features = false,  features = [ "r1cs" ] }
```
Comment (Pratyush) [inline Cargo.toml:25]: We should enable the `r1cs` feature only when we use it.

```suggestion
ark-crypto-primitives = { git = "https://github.com/arkworks-rs/crypto-primitives.git", branch = "main", default-features = false }
```

File: src/constraints.rs [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+    pub c: P::G1Var,
+}
+
+/// The verifying key variable for the Groth16 construction
+#[derive(Derivative)]
+#[derivative(
+    Clone(bound = "P::G1Var: Clone, P::GTVar: Clone, P::G1PreparedVar: Clone, \
+    P::G2PreparedVar: Clone, ")
+)]
+pub struct VerifyingKeyVar<E: PairingEngine, P: PairingVar<E>> {
```
Comment (Pratyush) [inline src/constraints.rs:35]:
```suggestion
/// A variable representing the Groth16 verifying key in the constraint system.
#[derive(Derivative)]
#[derivative(
    Clone(bound = "P::G1Var: Clone, P::GTVar: Clone, P::G1PreparedVar: Clone, \
    P::G2PreparedVar: Clone, ")
)]
pub struct VerifyingKeyVar<E: PairingEngine, P: PairingVar<E>> {
```

File: src/constraints.rs [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+    /// The `alpha * G`, where `G` is the generator of `E::G1`.
+    pub alpha_g1: P::G1Var,
+    /// The `beta * G`, where `G` is the generator of `E::G2`.
+    pub beta_g2: P::G2Var,
+    /// The `gamma * G`, where `G` is the generator of `E::G2`.
+    pub gamma_g2: P::G2Var,
+    /// The `delta * G`, where `G` is the generator of `E::G2`.
+    pub delta_g2: P::G2Var,
+    /// The `gamma^{-1} * (alpha * a_i + beta * b_i + c_i) * H`, where `H` is the generator of `E::G1`.
+    pub gamma_abc_g1: Vec<P::G1Var>,
```
Comment (Pratyush) [inline src/constraints.rs:45]: The existing comments don't really convey any extra information.

```suggestion
    #[doc(hidden)]
    pub alpha_g1: P::G1Var,
    #[doc(hidden)]
    pub beta_g2: P::G2Var,
    #[doc(hidden)]
    pub gamma_g2: P::G2Var,
    #[doc(hidden)]
    pub delta_g2: P::G2Var,
    #[doc(hidden)]
    pub gamma_abc_g1: Vec<P::G1Var>,
```

File: src/constraints.rs [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+}
+
+impl<E: PairingEngine, P: PairingVar<E>> VerifyingKeyVar<E, P> {
+    /// Prepare the verifying key `vk` for use in proof verification.
```
Comment (Pratyush) [inline src/constraints.rs:49]:
```suggestion
    /// Prepare `self` for use in proof verification.
```

File: src/constraints.rs [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+)]
+pub struct PreparedVerifyingKeyVar<E: PairingEngine, P: PairingVar<E>> {
+    /// The element `e(alpha * G, beta * H)` in `E::GT`.
+    pub alpha_g1_beta_g2: P::GTVar,
+    /// The element `- gamma * H` in `E::G2`, prepared for use in pairings.
+    pub gamma_g2_neg_pc: P::G2PreparedVar,
+    /// The element `- delta * H` in `E::G2`, prepared for use in pairings.
+    pub delta_g2_neg_pc: P::G2PreparedVar,
+    /// The `gamma^{-1} * (alpha * a_i + beta * b_i + c_i) * H`, where `H` is the generator of `E::G1`.
+    pub gamma_abc_g1: Vec<P::G1Var>,
```
Comment (Pratyush) [inline src/constraints.rs:81]: Shouldn't this also contain `vk`?

```suggestion
    #[doc(hidden)]
    pub alpha_g1_beta_g2: P::GTVar,
    #[doc(hidden)]
    pub gamma_g2_neg_pc: P::G2PreparedVar,
    #[doc(hidden)]
    pub delta_g2_neg_pc: P::G2PreparedVar,
    #[doc(hidden)]
    pub gamma_abc_g1: Vec<P::G1Var>,
```

File: src/constraints.rs [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+}
+
+/// The SNARK verifier gadget of [[Groth16]](https://eprint.iacr.org/2016/260.pdf).
```
Comment (Pratyush) [inline src/constraints.rs:84]:
```suggestion
/// Constraints for the verifier of the SNARK of [[Groth16]](https://eprint.iacr.org/2016/260.pdf).
```

File: Cargo.toml [hunk 7]
```diff
@@ -22,23 +22,27 @@ ark-poly = { git = "https://github.com/arkworks-rs/algebra", default-features =
...
+ark-r1cs-std = { git = "https://github.com/arkworks-rs/r1cs-std", default-features = false, optional = true }
+
+tracing = { version = "0.1", default-features = false, features = [ "attributes" ], optional = true }
+derivative = { version = "2.0", features = ["use_core"] }
```
Comment (Pratyush) [inline Cargo.toml:29]:
```suggestion
derivative = { version = "2.0", features = ["use_core"], optional = true}
```

File: Cargo.toml [hunk 8]
```diff
@@ -22,23 +22,27 @@ ark-poly = { git = "https://github.com/arkworks-rs/algebra", default-features =
...
+ark-mnt4-298 = { git = "https://github.com/arkworks-rs/curves", default-features = false, features = ["r1cs", "curve"] }
+ark-mnt6-298 = { git = "https://github.com/arkworks-rs/curves", default-features = false, features = ["r1cs"] }
 
 [features]
 default = ["parallel"]
-std = ["ark-ff/std", "ark-ec/std", "ark-poly/std", "ark-relations/std", "ark-std/std" ]
-parallel = ["std", "ark-ff/parallel", "ark-poly/parallel", "ark-ec/parallel", "ark-std/parallel", "rayon"]
+std = ["ark-ff/std", "ark-ec/std", "ark-poly/std", "ark-relations/std", "ark-crypto-primitives/std", "ark-std/std" ]
+parallel = ["std", "ark-ff/parallel", "ark-poly/parallel", "ark-ec/parallel", "ark-crypto-primitives/parallel", "ark-std/parallel", "rayon"]
+r1cs = [ "ark-crypto-primitives/r1cs", "ark-r1cs-std", "tracing" ]
```
Comment (Pratyush) [inline Cargo.toml:45]:
```suggestion
r1cs = [ "ark-crypto-primitives/r1cs", "ark-r1cs-std", "tracing", "derivative" ]
```

File: src/constraints.rs [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+            let c = cs.new_input_variable(|| {
+                let mut a = self.a.ok_or(SynthesisError::AssignmentMissing)?;
+                let b = self.b.ok_or(SynthesisError::AssignmentMissing)?;
+
+                a.mul_assign(&b);
+                Ok(a)
+            })?;
+
+            for _ in 0..(self.num_variables - 3) {
+                let _ =
```
Comment (npwardberkeley) [inline src/constraints.rs:441]: is the `let _ =` necessary?

File: src/constraints.rs [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+            let alpha_g1 =
+                P::G1Var::new_variable(ark_relations::ns!(cs, "alpha_g1"), || Ok(alpha_g1), mode)?;
+            let beta_g2 =
+                P::G2Var::new_variable(ark_relations::ns!(cs, "beta_g2"), || Ok(beta_g2), mode)?;
+            let gamma_g2 =
+                P::G2Var::new_variable(ark_relations::ns!(cs, "gamma_g2"), || Ok(gamma_g2), mode)?;
+            let delta_g2 =
+                P::G2Var::new_variable(ark_relations::ns!(cs, "delta_g2"), || Ok(delta_g2), mode)?;
+
+            let gamma_abc_g1 = Vec::new_variable(cs.clone(), || Ok(gamma_abc_g1), mode)?;
```
Comment (npwardberkeley) [inline src/constraints.rs:342]: I feel like the newline should be after this line. Also, is there a reason this doesn't get a namespace like the others do?

**Review by npwardberkeley — APPROVED** [review]
very minor comments

File: src/constraints.rs [hunk 11]
```diff
@@ -0,0 +1,536 @@
...
+    }
+
+    #[tracing::instrument(target = "r1cs", skip(circuit_vk, x, proof))]
+    fn verify(
+        circuit_vk: &Self::VerifyingKeyVar,
+        x: &Self::InputVar,
+        proof: &Self::ProofVar,
+    ) -> Result<Boolean<E::Fq>, SynthesisError> {
+        let pvk = circuit_vk.prepare()?;
+        Self::verify_with_processed_vk(&pvk, x, proof)
```
[the remark this answers is not in the document]
Comment (weikengchen) [reply src/constraints.rs:249]: Go for `process` according to https://github.com/arkworks-rs/snark/blob/master/snark/src/lib.rs

**Pull request merged**